File: 백준/Platinum/14003. 가장 긴 증가하는 부분 수열 5/가장 긴 증가하는 부분 수열 5.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Main {
    static StringTokenizer st;
    static int[] seq, dp;
    static List<Integer> lis;
    static int n;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        st = new StringTokenizer(br.readLine());
        seq = new int[n];

        for (int i = 0; i < n; i++) {
            seq[i] = nextInt();
        }

        dp = new int[n];
        lis = new ArrayList<>(n);

        for (int i = 0; i < n; i++) {
            int v = seq[i];

            if (lis.isEmpty() || v > lis.get(lis.size() - 1)) {
                lis.add(v);
                dp[i] = lis.size();
            } else {
                int idx = bisect(v);

                dp[i] = idx + 1;

                if (lis.get(idx) > v) {
                    lis.set(idx, v);
                }
            }
        }

        StringBuilder sb = new StringBuilder();

        int size = lis.size();
        int i = n;
        Deque<Integer> stack = new ArrayDeque<>(size);

        while (size > 0 && --i >= 0) {
            if (dp[i] == size) {
                stack.addLast(seq[i]);
                size--;
            }
        }

        sb.append(lis.size()).append('\n');

        while (!stack.isEmpty()) {
            sb.append(stack.removeLast()).append(' ');
        }

        System.out.println(sb);
        br.close();
    }

    static int bisect(int v) {
        int s = 0;
        int e = lis.size() - 1;
        int ret = 0;

        while (s <= e) {
            int mid = (s + e) >>> 1;
            int mv = lis.get(mid);

            if (mv == v) return mid;

            if (mv < v) {
                s = mid + 1;
            } else {
                ret = mid;
                e = mid - 1;
            }
        }

        return ret;
    }

    static int nextInt() {
        return Integer.parseInt(st.nextToken());
    }
}

File: 백준/Platinum/14003. 가장 긴 증가하는 부분 수열 5/가장 긴 증가하는 부분 수열 5.kt
fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val lis = LIS(n, readLine().split(' ').map(String::toInt))
    close()

    val bw = System.out.bufferedWriter()
    bw.append(lis.length.toString())
    bw.newLine()

    lis.sequence.joinTo(bw, " ")

    bw.flush()
    bw.close()
}

class LIS(n: Int, seq: List<Int>) {
    private val counts = IntArray(n) { 1 }
    private val buffer = arrayListOf(seq[0])
    private var longestIndex = 0
    val length get() = buffer.size
    val sequence = arrayListOf<Int>()

    init {
        for (i in 1 until n) {
            update(i, seq[i])
        }
        var count = length
        for (i in longestIndex downTo 0) {
            if (counts[i] == count) {
                sequence.add(seq[i])
                if (--count == 0) break
            }
        }
        sequence.reverse()
    }

    private fun update(i: Int, target: Int) {
        if (buffer[buffer.lastIndex] < target) {
            buffer.add(target)
            counts[i] = buffer.size
            longestIndex = i
        } else {
            val idx = bisect(target)
            if (buffer[idx] > target) {
                buffer[idx] = target
            }
            counts[i] += idx
        }
    }

    private fun bisect(t: Int): Int {
        var s = 0
        var e = buffer.lastIndex

        while (s < e) {
            val mid = (s + e) ushr 1
            val v = buffer[mid]

            if (v == t) return mid

            if (v < t) {
                s = mid + 1
            } else {
                e = mid
            }
        }
        return e
    }
}